ppcs/ppcTODO-signature-named-parameters.md

```diff
@@ -0,0 +1,153 @@
+# Named Parameters in Signatures
+
+## Preamble
+
+    Author:  Paul Evans <PEVANS>
+    Sponsor:
+    ID:      TODO
+    Status:  Exploratory
+
+## Abstract
+
+Adds the ability for subroutine signatures to process named arguments in the form of name-value pairs passed by the caller, in a fashion familiar to existing uses of assignment into a hash.
+
+## Motivation
+
+Perl 5.20 added "subroutine signatures", native syntax for more succinctly expressing the common patterns of processing arguments inbound into a subroutine by unpacking the `@_` array. Rather than writing common styles of code directly, a more compact notation, in a style immediately familiar to users of many other languages, allows Perl to create the required behaviour from that specification.
+
+```perl
+sub f ($x, $y, $z = 123) { ... }
+
+# instead of
+sub f {
+    die "Too many arguments" if @_ > 3;
+    die "Too few arguments" if @_ < 2;
+    my ($x, $y, $z) = @_;
+    $z = 123 if @_ < 3;
+    ...
+}
+```
+
+One common form of argument processing involves passing an even-sized list of key/value pairs and assigning that into a hash within the subroutine, so that specifically named parameters can be extracted from it. There is currently no support from subroutine signature syntax to assist authors in providing such behaviours.
+
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+A new kind of element may be present in a subroutine signature, which consumes a named argument from the caller. These elements are written with a leading colon prefix (`:$name`), indicating that it is named rather than positional. The name of each parameter is implied by the name of the lexical into which it is assigned, minus the leading `$` sigil.
+
+Each element provides a new lexical variable that is visible during the body of the function, in the same manner as positional ones.
+
+The value of a named parameter is taken from the argument values passed by the caller, in a manner familiar to existing uses of hash assignment. The caller should pass an even-sized name-value pair list. The values corresponding to names of parameters will be assigned into the variables. The order in which the values are passed by the caller is not significant.
+
+Furthemore, all of the new behaviour is performed within the body of the invoked subroutine entirely by inspecting the values of the arguments that were passed. The subroutine is not made aware of how those values came to be passed in - whether from literal name-value syntax, a hash or array variable expansion, or any other expression yielding such a list of argument name and value pairs.
+
+```perl
+sub make_colour ( :$red, :$green, :$blue ) { ... }
+
+make_colour( red => 1.0, blue => 0.5, green => 0.2 );
```

Comment:
let this syntax to use sigils, it may later allow you to use multiple lists
such syntax may also allow developer to use named syntax for positional parameters

Reply:
That's entirely against the whole point. The whole point as the document states, is that the callsites for these functions do not need to be changed or know about it. Much like with existing signature syntax, these extensions are simply better ways that the *body* of the function could be implemented, without needing the caller of the function to know about it.

Reply:
hm, you are most likely right, with existing code and current state of syntax and parser it may be difficult to determine state I'd like to have without prefix token. FYI, I was thinking about function like one in `Net::Amazon::S3`, for example `add_key`, which you can call as:
````
add_key ($bucket, $key);
add_key (bucket => $bucket, key => $key);
add_key ($key, bucket => $bucket);
add_key ($key, name => $bucket); # named parameter bucket has an alias
```

Reply:
we had discussed this on the mailing list many moons ago, that allowing users to use named syntax for positional parameters is not good behavior by default b/c it suddenly exposes your internal API as external (b/c the names you happen to have chosen as params now matter at the call site).

Reply:
> we had discussed this on the mailing list many moons ago, that allowing users to use named syntax for positional parameters is not good behavior by default b/c it suddenly exposes your internal API as external (b/c the names you happen to have chosen as params now matter at the call site).

Is this an argument unique to the implementation of named params here, or are you arguing that named params in the abstract are a bad idea?

Reply:
> > we had discussed this on the mailing list many moons ago, 
> 
> Is this an argument unique to the implementation of named params here, or are you arguing that named params in the abstract are a bad idea?

The argument I made was against the global call-site syntax that happy-barney recommended for named parameters. That has the issue of exposing your internal API, as mentioned. Named params in general are fine if you decide that's what your API should look like.

Reply:
@rabbiveesh 

regarding: positional vs name call syntax. That definitely should be up to caller. Most likely caller is aware of context where function is called therefor can adjust call syntax accordingly. Function author should not dictate how to call it, only what constraint applies.

Reply:
This kind of thing would be better handled via `multisub`s, no? That's kind of what `Type::Params` [allows](https://metacpan.org/release/TOBYINK/Type-Tiny-2.004000/view/lib/Type/Params.pm#multiple-ArrayRef), but I don't think that is good to have in a general case.